examples/step-64/step-64.cu

```diff
@@ -0,0 +1,576 @@
+    LinearAlgebra::distributed::Vector<double, MemorySpace::CUDA> &      dst,
+    const LinearAlgebra::distributed::Vector<double, MemorySpace::CUDA> &src)
+    const
+  {
+    dst = 0.;
+    LocalHelmholtzOperator<dim, fe_degree> helmholtz_operator(
+      coef.get_values());
+    mf_data.cell_loop(helmholtz_operator, src, dst);
+    mf_data.copy_constrained_values(src, dst);
+  }
+
+
+  // This class defines the usual framework we use for tutorial programs. The
+  // only point worth commenting on the solve() function and the choice of
+  // vector types.
+  template <int dim, int fe_degree>
+  class HelmholtzProblem
+  {
+  public:
+    HelmholtzProblem();
+    ~HelmholtzProblem();
+
+    void run();
+
+  private:
+    void setup_system();
+
+    void assemble_rhs();
+
+    void solve();
+
+    void output_results(const unsigned int cycle) const;
+
+    MPI_Comm mpi_communicator;
+
+    parallel::distributed::Triangulation<dim> triangulation;
+
+    DoFHandler<dim> dof_handler;
+    FE_Q<dim>       fe;
+
+    IndexSet locally_owned_dofs;
+    IndexSet locally_relevant_dofs;
+
+    AffineConstraints<double>                          constraints;
+    std::unique_ptr<HelmholtzOperator<dim, fe_degree>> system_matrix_dev;
+
+    // Since all the operations in the solve functions are executed on the
+    // graphic card it is necessary for the vectors used to store their values
+    // on the GPU as well. LinearAlgebra::distributed::Vector can be told which
+    // memory space to use. There is also LinearAlgebra::CUDAWrappers::Vector
+    // that always uses GPU memory storage but doesn't work with MPI. It might
+    // be worth noticing that the communication between different MPI processes
+    // can be improved if the MPI implementation is CUDA-aware and the configure
+    // flag DEAL_II_WITH_CUDA_AWARE_MPI is enabled.
+    //
+    // Here, we also have a finite element vector with CPU storage such that we
+    // can view and display the solution as usual.
+    LinearAlgebra::distributed::Vector<double, MemorySpace::Host>
+                                                                  ghost_solution_host;
+    LinearAlgebra::distributed::Vector<double, MemorySpace::CUDA> solution_dev;
+    LinearAlgebra::distributed::Vector<double, MemorySpace::CUDA>
+      system_rhs_dev;
+
+    ConditionalOStream pcout;
+  };
+
+
+  // The implementation of all the remaining functions of this class apart from
+  // Helmholtzproblem::solve() doesn't contain anything new and we won't further
+  // comment on it.
+  template <int dim, int fe_degree>
+  HelmholtzProblem<dim, fe_degree>::HelmholtzProblem()
+    : mpi_communicator(MPI_COMM_WORLD)
+    , triangulation(mpi_communicator)
+    , dof_handler(triangulation)
+    , fe(fe_degree)
+    , pcout(std::cout, Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+  {}
+
+
+
+  template <int dim, int fe_degree>
+  HelmholtzProblem<dim, fe_degree>::~HelmholtzProblem()
+  {
+    dof_handler.clear();
+  }
```

Comment:
I guess you can remove this destructor. To be sure, move `FE_Q` above `DoFHandler` in the `HelmholtzProblem` class declaration, even though we nowadays clone the FE rather than having the `SmartPointer` solution that required this step.

Reply:
Good catch!